model/task/task.go

```diff
@@ -2393,31 +2390,66 @@ func (t *Task) IsPartOfSingleHostTaskGroup() bool {
 }
 
 func (t *Task) IsPartOfDisplay() bool {
-	dt, err := t.GetDisplayTask()
-	if err != nil {
-		grip.Error(message.WrapError(err, message.Fields{
-			"message":        "unable to get display task",
-			"execution_task": t.Id,
-		}))
-		return false
+	// if display task ID is nil, we need to check manually if we have an execution task
+	if t.DisplayTaskId == nil {
+		dt, err := t.GetDisplayTask()
+		if err != nil {
+			grip.Error(message.WrapError(err, message.Fields{
+				"message":        "unable to get display task",
+				"execution_task": t.Id,
+			}))
+			return false
+		}
+		return dt != nil
 	}
-	return dt != nil
+
+	return utility.FromStringPtr(t.DisplayTaskId) != ""
 }
 
 func (t *Task) GetDisplayTask() (*Task, error) {
 	if t.DisplayTask != nil {
 		return t.DisplayTask, nil
 	}
+	dtId := utility.FromStringPtr(t.DisplayTaskId)
+	if t.DisplayTaskId != nil && dtId == "" {
+		// display task ID is explicitly set to empty if it's not a display task
+		return nil, nil
+	}
 	var dt *Task
 	var err error
 	if t.Archived {
-		dt, err = FindOneOld(ByExecutionTask(t.OldTaskId))
+		if dtId != "" {
+			dt, err = FindOneOldNoMergeByIdAndExecution(dtId, t.Execution)
+		} else {
+			dt, err = FindOneOld(ByExecutionTask(t.OldTaskId))
+			if dt != nil {
+				dtId = dt.OldTaskId // save the original task ID to cache
+			}
+		}
 	} else {
-		dt, err = FindOne(ByExecutionTask(t.Id))
+		if dtId != "" {
+			dt, err = FindOneId(dtId)
+		} else {
+			dt, err = FindOne(ByExecutionTask(t.Id))
+			if dt != nil {
+				dtId = dt.Id
+			}
+		}
 	}
 	if err != nil {
 		return nil, err
 	}
+
+	if t.DisplayTaskId == nil {
+		// Cache display task ID for future use. If we couldn't find the display task,
+		// we cache the empty string to show that it doesn't exist.
```

Comment:
So does and empty string reliably mean that it doesn't exist and is therefor not an execution task?  If so, what did you mean by `for many tasks it will be populated but empty`? Are there cases where it's empty but doesn't reliably mean that it's not an execution task?

Reply:
Yes, that is the behavior -- by "It will be populated but empty" I meant that it will be set as a pointer to an empty string. The case where it's nil is the case that isn't reliable.